numpy/_core/src/umath/dispatching.c

```diff
@@ -889,6 +889,35 @@ promote_and_get_info_and_ufuncimpl(PyUFuncObject *ufunc,
     return info;
 }
 
+/*
+ * Fast path for promote_and_get_info_and_ufuncimpl.
+ * Enters a critical section if there is a cache miss.
+ */
+static inline PyObject *
+try_promote_and_get_info_and_ufuncimpl(PyUFuncObject *ufunc,
+        PyArrayObject *const ops[],
+        PyArray_DTypeMeta *signature[],
+        PyArray_DTypeMeta *op_dtypes[],
+        npy_bool legacy_promotion_is_possible)
+{
+    PyObject *info = PyArrayIdentityHash_GetItem(ufunc->_dispatch_cache,
```

Comment:
I don't think this is thread-safe because we may read from the hash table while another thread is concurrently modifying it, including possibly resizing the table.

Reply:
We could make PyArrayIdentityHash a concurrent hash table. That's generally hard, but there are a few things that make this specific case easier:

* We only care about concurrent read performance
* Entries are never deleted or overwritten

But we'd still have to deal with hash table resizing, which makes it more complex.

I think a RW lock will be simpler as long as it provides good enough read performance.